src/Elastic.Transport/Components/Serialization/SystemTextJsonSerializer.cs

```diff
@@ -0,0 +1,168 @@
+
+		if (TryReturnDefault(stream, out T deserialize))
+			return deserialize;
+
+		return JsonSerializer.Deserialize<T>(stream, _options);
+	}
+
+	/// <inheritdoc />
+	public override object? Deserialize(Type type, Stream stream)
+	{
+		Initialize();
+
+		if (TryReturnDefault(stream, out object deserialize))
+			return deserialize;
+
+		return JsonSerializer.Deserialize(stream, type, _options);
+	}
+
+	/// <inheritdoc />
+	public override ValueTask<T> DeserializeAsync<T>(Stream stream, CancellationToken cancellationToken = default)
+	{
+		Initialize();
+
+		if (TryReturnDefault(stream, out T deserialize))
+			return new ValueTask<T>(deserialize);
+
+		return JsonSerializer.DeserializeAsync<T>(stream, _options, cancellationToken);
+	}
+
+	/// <inheritdoc />
+	public override ValueTask<object?> DeserializeAsync(Type type, Stream stream, CancellationToken cancellationToken = default)
+	{
+		Initialize();
+
+		if (TryReturnDefault(stream, out object deserialize))
+			return new ValueTask<object?>(deserialize);
+
+		return JsonSerializer.DeserializeAsync(stream, type, _options, cancellationToken);
+	}
+
+	/// <inheritdoc />
+	public override void Serialize<T>(T data, Stream writableStream,
+		SerializationFormatting formatting = SerializationFormatting.None)
+	{
+		Initialize();
+
+		JsonSerializer.Serialize(writableStream, data, GetJsonSerializerOptions(formatting));
+	}
+
+	/// <inheritdoc />
+	public override Task SerializeAsync<T>(T data, Stream stream,
+		SerializationFormatting formatting = SerializationFormatting.None,
+		CancellationToken cancellationToken = default)
+	{
+		Initialize();
+
+		return JsonSerializer.SerializeAsync(stream, data, GetJsonSerializerOptions(formatting), cancellationToken);
+	}
+
+	#endregion Serializer
+
+	/// <summary>
+	/// A factory method that can create an instance of <see cref="JsonSerializerOptions"/> that will
+	/// be used when serializing.
+	/// </summary>
+	/// <returns></returns>
+	protected abstract JsonSerializerOptions? CreateJsonSerializerOptions();
+
+	/// <summary>
+	/// A callback function that is invoked after the <see cref="JsonSerializerOptions"/> have been created and the
+	/// serializer got fully initialized.
+	/// </summary>
+	protected virtual void Initialized()
+	{
+	}
+
+	/// <summary>
+	/// Returns the <see cref="JsonSerializerOptions"/> for this serializer, based on the given <paramref name="formatting"/>.
+	/// </summary>
+	/// <param name="formatting">The serialization formatting.</param>
+	/// <returns>The requested <see cref="JsonSerializerOptions"/> or <c>null</c>, if the serializer is not initialized yet.</returns>
+	protected internal JsonSerializerOptions? GetJsonSerializerOptions(SerializationFormatting formatting) => (formatting is SerializationFormatting.None)
+		? _options
+		: _indentedOptions;
+
+	/// <summary>
+	/// Initializes a serializer instance such that its <see cref="JsonSerializerOptions"/> are populated.
+	/// </summary>
+	protected internal void Initialize()
```

Comment:
This lazy initialization style is copied from the `SystemTextJsonSerializer` implementation which is currently present in `Elastic.Clients.Elasticsearch` and improved to make it thread-safe.

The `SemaphoreSlim` causes an extra allocation, but compared to the old `LowLevelRequestResponseSerializer` implementation, this saves the allocation of 2x `Lazy` instances.